## City Bike

### 1. Import data
Timespan: July 2017 - July 2018

In [1]:
import pandas as pd

In [2]:
df201807 = pd.read_csv("JC-201807-citibike-tripdata.csv")
df201806 = pd.read_csv("JC-201806-citibike-tripdata.csv")
df201805 = pd.read_csv("JC-201805-citibike-tripdata.csv")
df201804 = pd.read_csv("JC-201804-citibike-tripdata.csv")
df201803 = pd.read_csv("JC-201803-citibike-tripdata.csv")
df201802 = pd.read_csv("JC-201802-citibike-tripdata.csv")
df201801 = pd.read_csv("JC-201801-citibike-tripdata.csv")
df201712 = pd.read_csv("JC-201712-citibike-tripdata.csv")
df201711 = pd.read_csv("JC-201711-citibike-tripdata.csv")
df201710 = pd.read_csv("JC-201710-citibike-tripdata.csv")
df201709 = pd.read_csv("JC-201709-citibike-tripdata.csv")
df201708 = pd.read_csv("JC-201708-citibike-tripdata.csv")
df201707 = pd.read_csv("JC-201707-citibike-tripdata.csv")
df201807.head()

tripduration                 starttime                  stoptime  \
0          1234  2018-07-01 02:08:39.8920  2018-07-01 02:29:14.4890   
1          1041  2018-07-01 02:09:07.5180  2018-07-01 02:26:28.8720   
2          2308  2018-07-01 02:14:45.7400  2018-07-01 02:53:14.5970   
3           767  2018-07-01 02:25:12.4020  2018-07-01 02:38:00.0510   
4           151  2018-07-01 04:07:44.0520  2018-07-01 04:10:15.8010   

   start station id start station name  start station latitude  \
0              3183     Exchange Place               40.716247   
1              3183     Exchange Place               40.716247   
2              3183     Exchange Place               40.716247   
3              3183     Exchange Place               40.716247   
4              3183     Exchange Place               40.716247   

   start station longitude  end station id end station name  \
0               -74.033459            3183   Exchange Place   
1               -74.033459            3183   Exchange Place   
2               -74.033459            3183   Exchange Place   
3               -74.033459            3183   Exchange Place   
4               -74.033459            3275   Columbus Drive   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.716247             -74.033459   26306    Customer   
1             40.716247             -74.033459   29204    Customer   
2             40.716247             -74.033459   26234    Customer   
3             40.716247             -74.033459   26238  Subscriber   
4             40.718355             -74.038914   29204  Subscriber   

   birth year  gender  
0        1969       0  
1        1969       0  
2        1969       0  
3        1992       1  
4        1972       1

In [3]:
df201707.head()

tripduration            starttime             stoptime  start station id  \
0           340  2017-07-01 00:00:08  2017-07-01 00:05:48              3186   
1           439  2017-07-01 00:02:35  2017-07-01 00:09:55              3183   
2           186  2017-07-01 00:04:33  2017-07-01 00:07:40              3186   
3           442  2017-07-01 00:05:53  2017-07-01 00:13:15              3195   
4           170  2017-07-01 00:07:54  2017-07-01 00:10:45              3187   

  start station name  start station latitude  start station longitude  \
0      Grove St PATH               40.719586               -74.043117   
1     Exchange Place               40.716247               -74.033459   
2      Grove St PATH               40.719586               -74.043117   
3            Sip Ave               40.730743               -74.063784   
4          Warren St               40.721124               -74.038051   

   end station id end station name  end station latitude  \
0            3267     Morris Canal             40.712419   
1            3199     Newport Pkwy             40.728745   
2            3270  Jersey & 6th St             40.725289   
3            3193     Lincoln Park             40.724605   
4            3186    Grove St PATH             40.719586   

   end station longitude  bikeid    usertype  birth year  gender  
0             -74.038526   29643  Subscriber      1994.0       2  
1             -74.032108   29585  Subscriber      1980.0       1  
2             -74.045572   29219  Subscriber      1984.0       1  
3             -74.078406   29497  Subscriber      1969.0       1  
4             -74.043117   29483  Subscriber      1986.0       1

In [4]:
# Concatenate JC dataframes
df = pd.concat([df201707, df201708, df201709, df201710, df201711, df201712, df201801, \
           df201802, df201803, df201804, df201805, df201806, df201807])
df.head()

tripduration            starttime             stoptime  start station id  \
0           340  2017-07-01 00:00:08  2017-07-01 00:05:48              3186   
1           439  2017-07-01 00:02:35  2017-07-01 00:09:55              3183   
2           186  2017-07-01 00:04:33  2017-07-01 00:07:40              3186   
3           442  2017-07-01 00:05:53  2017-07-01 00:13:15              3195   
4           170  2017-07-01 00:07:54  2017-07-01 00:10:45              3187   

  start station name  start station latitude  start station longitude  \
0      Grove St PATH               40.719586               -74.043117   
1     Exchange Place               40.716247               -74.033459   
2      Grove St PATH               40.719586               -74.043117   
3            Sip Ave               40.730743               -74.063784   
4          Warren St               40.721124               -74.038051   

   end station id end station name  end station latitude  \
0            3267     Morris Canal             40.712419   
1            3199     Newport Pkwy             40.728745   
2            3270  Jersey & 6th St             40.725289   
3            3193     Lincoln Park             40.724605   
4            3186    Grove St PATH             40.719586   

   end station longitude  bikeid    usertype  birth year  gender  
0             -74.038526   29643  Subscriber      1994.0       2  
1             -74.032108   29585  Subscriber      1980.0       1  
2             -74.045572   29219  Subscriber      1984.0       1  
3             -74.078406   29497  Subscriber      1969.0       1  
4             -74.043117   29483  Subscriber      1986.0       1

### 2. Data exploration

In [5]:
df.describe()

tripduration  start station id  start station latitude  \
count  3.627480e+05     362748.000000           362748.000000   
mean   6.913486e+02       3242.878596               40.722536   
std    7.997063e+03        107.912237                0.007367   
min    6.100000e+01       3183.000000               40.692640   
25%    2.320000e+02       3187.000000               40.717733   
50%    3.400000e+02       3203.000000               40.721525   
75%    5.700000e+02       3269.000000               40.727224   
max    2.181628e+06       3694.000000               40.752559   

       start station longitude  end station id  end station latitude  \
count            362748.000000   362748.000000         362748.000000   
mean                -74.045865     3238.477472             40.634089   
std                   0.010576      121.629334              1.891122   
min                 -74.096937       79.000000              0.000000   
25%                 -74.050389     3186.000000             40.716540   
50%                 -74.043845     3203.000000             40.719586   
75%                 -74.038051     3267.000000             40.727224   
max                 -74.032108     3694.000000             40.814326   

       end station longitude         bikeid     birth year         gender  
count          362748.000000  362748.000000  351280.000000  362748.000000  
mean              -73.885302   29437.154700    1979.936615       1.141961  
std                 3.438635    2427.668509      10.291742       0.507825  
min               -74.096937   14529.000000    1887.000000       0.000000  
25%               -74.049968   26308.000000    1974.000000       1.000000  
50%               -74.043117   29527.000000    1982.000000       1.000000  
75%               -74.038051   31715.000000    1988.000000       1.000000  
max                 0.000000   33695.000000    2002.000000       2.000000

362748 trips have been recorded total during the chosen period.

In [6]:
pctg = round((len(df201807) / len(df201707) - 1)*100, 1)
print(f"From 2017-08 to 2018-08, total ridership has grown by {pctg}%.")

From 2017-08 to 2018-08, total ridership has grown by 25.9%.


In [7]:
sub201807 = df201807["usertype"].value_counts()["Subscriber"]
print("2018-07")
print(f"Proportion of subscribers: {sub201807 / df201807['usertype'].count()}")
cus201807 = df201807["usertype"].value_counts()["Customer"]
print(f"Proportion of customers: {cus201807 / df201807['usertype'].count()}")

sub201707 = df201707["usertype"].value_counts()["Subscriber"]
print("---\n2017-07")
print(f"Proportion of subscribers: {sub201707 / df201707['usertype'].count()}")
cus201707 = df201707["usertype"].value_counts()["Customer"]
print(f"Proportion of customers: {cus201707 / df201707['usertype'].count()}")

2018-07
Proportion of subscribers: 0.9102394246238289
Proportion of customers: 0.0897605753761711
---
2017-07
Proportion of subscribers: 0.9217228129747118
Proportion of customers: 0.07827718702528817


In [8]:
summer = pd.concat([df201806, df201807, df201708])
summer["hour"] = pd.to_datetime(summer.starttime)
summer["hour"] = summer["hour"].dt.hour
summer.groupby(["hour"])["bikeid"].count()

hour
0       985
1       469
2       240
3       156
4       248
5      1097
6      3379
7      7712
8     13567
9      7486
10     4898
11     4609
12     5070
13     5041
14     4902
15     4920
16     6063
17    10660
18    12061
19     9381
20     6734
21     4424
22     2923
23     1652
Name: bikeid, dtype: int64

Peak hours are 8:00, 17:00, and 18:00 during summer months June, July, August.

In [9]:
winter = pd.concat([df201712, df201801, df201802])
winter["hour"] = pd.to_datetime(winter.starttime)
winter["hour"] = winter["hour"].dt.hour
winter.groupby(["hour"])["bikeid"].count()

hour
0      355
1      227
2      108
3       69
4      120
5      586
6     1410
7     3144
8     5128
9     2677
10    1676
11    1477
12    1878
13    1882
14    1871
15    1987
16    2393
17    4227
18    4696
19    2947
20    2067
21    1378
22     849
23     527
Name: bikeid, dtype: int64

Peak hours are 8:00, 17:00, and 18:00 during winter months December, January, Febuary.

#### Today, top 10 stations in the city for starting a journey

In [10]:
# Query today's date from the dataframe
d='2017-09-09' 
totay_date=pd.to_datetime(d).date()

df201709["date"] = pd.to_datetime(df201709["starttime"])
df201709["date"] = df201709["date"].dt.date
df_queried_date = df201709.loc[df201709["date"] == totay_date]

In [11]:
# Find top stations
df_top_stations = df_queried_date.groupby(["start station id"])\
    ["start station id"].count().sort_values(ascending=False)[0:10].to_frame()
df_top_stations = df_top_stations.rename(columns={"start station id":"number of start stations"})
df_top_stations["start station id"] = df_top_stations.index
df_top_stations.reset_index(drop=True, inplace=True)
# df_top_stations

In [12]:
# Report top station names
stations = df201709[["start station id", "start station name"]].drop_duplicates()
df_top_stations.merge(stations, how="inner", on="start station id")

number of start stations  start station id start station name
0                        79              3186      Grove St PATH
1                        55              3183     Exchange Place
2                        53              3203      Hamilton Park
3                        50              3267       Morris Canal
4                        42              3276   Marin Light Rail
5                        41              3211         Newark Ave
6                        40              3202       Newport PATH
7                        40              3195            Sip Ave
8                        34              3199       Newport Pkwy
9                        29              3272       Jersey & 3rd

#### Today, top 10 stations in the city for ending a journe

In [13]:
# Find top stations
df_top_stations = df_queried_date.groupby(["end station id"])\
    ["end station id"].count().sort_values(ascending=False)[0:10].to_frame()
df_top_stations = df_top_stations.rename(columns={"end station id":"number of end stations"})
df_top_stations["end station id"] = df_top_stations.index
df_top_stations.reset_index(drop=True, inplace=True)
# df_top_stations

In [14]:
# Report top station names
stations = df201709[["end station id", "end station name"]].drop_duplicates()
df_top_stations.merge(stations, how="inner", on="end station id")

number of end stations  end station id  end station name
0                     121            3186     Grove St PATH
1                      53            3203     Hamilton Park
2                      52            3211        Newark Ave
3                      46            3183    Exchange Place
4                      45            3267      Morris Canal
5                      43            3195           Sip Ave
6                      40            3202      Newport PATH
7                      37            3276  Marin Light Rail
8                      34            3187         Warren St
9                      33            3199      Newport Pkwy

### 3. Save combined data

In [15]:
df.to_csv("past_year.csv")